networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,577 @@
+        } while (!done);
+        flattenPartition();
+        PLPRINT("Leiden done, took " << totalTime.elapsedTag() << " Modularity: " << setprecision(8)
+                                     << Modularity().getQuality(result, *G))
+    } while (changed && numberOfIterations > 0);
+    hasRun = true;
+}
+
+void ParallelLeiden::calculateVolumes(const Graph &graph) {
+#ifdef PLFINE
+    auto timer = Aux::Timer();
+    timer.start();
+#endif
+    // thread safe reduction. Avoid atomic calculation of total graph volume for unweighted graphs.
+    // Vol(G) is then 2*|E|
+    communityVolumes.clear();
+    communityVolumes.resize(result.upperBound() + 10000);
+    if (graph.isWeighted()) {
+        std::vector<double> threadVolumes(omp_get_max_threads());
+        graph.parallelForNodes([&](node a) {
+            {
+                edgeweight ew = graph.weightedDegree(a, true);
+#pragma omp atomic
+                communityVolumes[result[a]] += ew;
+                threadVolumes[omp_get_thread_num()] += ew;
+            }
+        });
+        for (const auto vol : threadVolumes) {
+            inverseGraphVolume += vol;
+        }
+        inverseGraphVolume = 1 / inverseGraphVolume;
+    } else {
+        inverseGraphVolume = 1.0 / (2 * graph.numberOfEdges());
+        graph.parallelForNodes([&](node a) {
+            {
+#pragma omp atomic
+                communityVolumes[result[a]] += graph.weightedDegree(a, true);
+            }
+        });
+    }
+    PLTRACE("Calculating Volumes took " << timer.elapsedMilliseconds() / 1000.0 << "s")
+}
+
+void ParallelLeiden::flattenPartition() {
+#ifdef PLFINE
+    auto timer = Aux::Timer();
+    timer.start();
+#endif
+    if (mappings.empty()) {
+        return;
+    }
+    Partition flattenedPartition(
+        G->numberOfNodes()); // Create a new partition with size |V(G)| (the fine/bigger Graph)
+    flattenedPartition.setUpperBound(result.upperBound());
+    int i = mappings.size() - 1;
+    std::vector<node> &lower = mappings[i--];
+    while (i >= 0) { // iteratively "resolve" (i.e compose) mappings. Let "lower" be a mapping thats
+                     // below "higher" in the hierarchy (i.e. of a later aggregation)
+        std::vector<node> &upper =
+            mappings[i--]; // If higher[index] = z and lower[z] = x then: higher[index] = x
+        for (auto &idx : upper) {
+            idx = lower[idx];
+        }
+        lower = upper;
+    }
+    G->parallelForNodes([&](node a) { flattenedPartition[a] = lower[a]; });
+    flattenedPartition.compact(true);
+    result = flattenedPartition;
+    mappings.clear();
+    PLTRACE("Flattening partition took " << timer.elapsedMilliseconds() / 1000.0 << "s")
+}
+
+void ParallelLeiden::parallelMove(const Graph &graph) {
+#ifdef PLFINE
+    cout << "Local Moving : " << graph.numberOfNodes() << " Nodes " << endl;
+    std::vector<long> moved(omp_get_max_threads(), 0);
+    std::vector<long> totalNodesPerThread(omp_get_max_threads(), 0);
```

Comment:
Instead of using `long` (signed datatype) as vector-items, better use `count` (alias for `uint64_t`) to avoid overflow for huge number of nodes.  In addition -  if I understand your code correctly, the items are always positive anyways.